drivers: flash: flash_stm32g4x.c: manage bank1/2 discontinuity

ABOSTM opened the pull request:
drivers: flash: flash_stm32g4x.c: manage bank1/2 discontinuity

When flash is Dualbank and flash size is lower than 512K,
then there is a discontinutity between bank1 and bank2.

Fixes #30148

ABOSTM committed:
drivers: flash: flash_stm32g4x.c: manage bank1/2 discontinuity

When flash is Dualbank and flash size is lower than 512K,
then there is a discontinuity between bank1 and bank2.

Signed-off-by: Alexandre Bourdiol <alexandre.bourdiol@st.com>

ABOSTM commented:
@Laczen @nvlsianpu, with your flash experience, can you please comment this proposal to manage discontinuity between bank1 and bank2 ?

azeemshatp commented:
Just tested this code on G474RCT and observed the below thing.

When I flash the code in Bank 2 (0x08040000) and write to the offset 0x5F800, it fails occasionally giving me a flash_write failure (err:-5). This happens only from Bank 2.

I also tested few other cases, which seems to be working fine. So,

App in Bank 2 (0x08040000),  write to 0x5F800 => Gives the above issue
App in Bank 2 (0x08040000),  write to 0x1F800 => OK
App in Bank 1 (0x08000000),  write to 0x5F800 => OK
App in Bank 1 (0x08000000),  write to 0x1F800 => OK

ABOSTM committed:
drivers: flash: flash_stm32g4x.c: rework flash_stm32_block_erase_loop

Remove hardcode bank1 page limit (128) which is not always valid
(not valid for soc stm32g474rct)
Manage bank1/2 discontinuity when flash is Dualbank and
flash size is lower than 512K.

Signed-off-by: Alexandre Bourdiol <alexandre.bourdiol@st.com>

ABOSTM commented:
Hi @azeemshatp,
I could not reproduce your issue because I don't have soc stm32g474rct (I only have nucleo_g474re) and because I face difficulties to run application on 2nd bank.
Nevertheless I found some hardcoded page limit value in erase code, which for sure don't match with smaller flash size like you have on stm32g474rct.
So I pushed a 2nd commit on this PR.
Even if I didn't found the link with the fact that you face issue only when erasing this specific address 0x5F800 from bank2,
I still hope this fixes the issue.
It works on my nucleo_g474re, but I couldn't test on stm32g474rct.
Could you test this new commit ? 
I would be interested by tests at address:
* 0x0805F800 from bank2 (the case which cause issue)
* 0x08020000 should failed as invalid address (between bank1 and bank2)

As erase operation is concerned, I propose you to update tests as follow, for each address
```
flash_erase();
flash_write();
flash_read();  // so that we can check write operation is effective
flash_erase();
flash_read(); // so that we can check erase operation is effective
```

Note: in case this doesn't fix your problem can you please enable switch FLASH_LOG_LEVEL_DBG and provide the log.

azeemshatp commented:
Doesn't seem to be fixed yet. Booted from Bank 2, with FLASH_LOG_LEVEL_DBG enabled for both the offsets mentioned.

0x08020000 works fine as below

D: Flash initialized. BS: 8
D: Block 0: bs: 2048 count: 64
D: Block 1: bs: 131072 count: 1
D: Block 2: bs: 2048 count: 64
D: Disable write protection
*** Booting Zephyr OS build zephyr-v2.4.0  ***
Found flash controller FLASH_CTRL.
Flash I/O commands can be run.
Flash offset: 20000 
D: Disable write protection
E: Erase range invalid. Offset: 131072, len: 64
flash_erase failed (err:-22).

However, a write to offset 0x5F800 doesn't work and I think the erasing to this offset is not actually happening correctly except at the beginning. I am unable to write again to the offset because the flash is not erased, even though the API returned a successful erase.

D: Flash initialized. BS: 8
D: Block 0: bs: 2048 count: 64
D: Block 1: bs: 131072 count: 1
D: Block 2: bs: 2048 count: 64
D: Disable write protection
*** Booting Zephyr OS build zephyr-v2.4.0  ***
Found flash controller FLASH_CTRL.
Flash I/O commands can be run.
Flash offset: 5F800 
D: Disable write protection
D: Erase offset: 391168, len: 64
D: Erase page 63 on bank 2
Reading 1 
D: Read offset: 391168, len: 64
Reading from flash .. 
ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
D: Write offset: 391168, len: 64
Reading 2 
D: Read offset: 391168, len: 64
Reading from flash .. 
00 01 02 03 04 05 06 07 08 09 0a 0b 0c 0d 0e 0f 10 11 12 13 14 15 16 17 18 19 1a 1b 1c 1d 1e 1f 20 21 22 23 24 25 26 27 28 29 2a 2b 2c 2d 2e 2f 30 31 32 33 34 35 36 37 38 39 3a 3b 3c 3d 3e 3f 
Erasing again... 
D: Erase offset: 391168, len: 64
D: Erase page 63 on bank 2
Reading 3 
D: Read offset: 391168, len: 64
Reading from flash .. 
00 01 02 03 04 05 06 07 08 09 0a 0b 0c 0d 0e 0f 10 11 12 13 14 15 16 17 18 19 1a 1b 1c 1d 1e 1f 20 21 22 23 24 25 26 27 28 29 2a 2b 2c 2d 2e 2f 30 31 32 33 34 35 36 37 38 39 3a 3b 3c 3d 3e 3f 
D: Enable write protection

ABOSTM commented:
@azeemshatp,
Can you please make complementary tests:
* the same (with my last commit) on 0x0805F800 but from bank1
* check if any other address in bank2 is working fine when app is in bank 2
ex : 0x0805F000 , or 0x08050000   (just take care to not erase the app)

ABOSTM commented:
@azeemshatp,
When you compile the app to run from bank2, can you tell me the value of switch CONFIG_FLASH_BASE_ADDRESS ?
0x8000000 or  0x8040000 ?

azeemshatp commented:
> @azeemshatp,
> Can you please make complementary tests:
> 
> * the same (with my last commit) on 0x0805F800 but from bank1

From Bank 1 using your commit:
It works fine for all addresses as I reported initially, even for writing to Bank 1 (0x1F800) itself.

D: Flash initialized. BS: 8
D: Block 0: bs: 2048 count: 64
D: Block 1: bs: 131072 count: 1
D: Block 2: bs: 2048 count: 64
D: Disable write protection
*** Booting Zephyr OS build zephyr-v2.4.0  ***
Found flash controller FLASH_CTRL.
Flash I/O commands can be run.
Flash offset: 5F800 
D: Disable write protection
D: Erase offset: 391168, len: 64
D: Erase page 63 on bank 2
Reading 1 
D: Read offset: 391168, len: 64
Reading from flash .. 
ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
D: Write offset: 391168, len: 64
Reading 2 
D: Read offset: 391168, len: 64
Reading from flash .. 
00 01 02 03 04 05 06 07 08 09 0a 0b 0c 0d 0e 0f 10 11 12 13 14 15 16 17 18 19 1a 1b 1c 1d 1e 1f 20 21 22 23 24 25 26 27 28 29 2a 2b 2c 2d 2e 2f 30 31 32 33 34 35 36 37 38 39 3a 3b 3c 3d 3e 3f 
Erasing again... 
D: Erase offset: 391168, len: 64
D: Erase page 63 on bank 2
Reading 3 
D: Read offset: 391168, len: 64
Reading from flash .. 
ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
D: Enable write protection


> * check if any other address in bank2 is working fine when app is in bank 2
>   ex : 0x0805F000 , or 0x08050000   (just take care to not erase the app)

It doesn't work for offset: 0x5F000 also. I should have checked this before, sorry for that. 
So the issue is not specifically to the last flash page, instead it could be for a write to the same bank while in Bank 2.

D: Flash initialized. BS: 8
D: Block 0: bs: 2048 count: 64
D: Block 1: bs: 131072 count: 1
D: Block 2: bs: 2048 count: 64
D: Disable write protection
*** Booting Zephyr OS build zephyr-v2.4.0  ***
Found flash controller FLASH_CTRL.
Flash I/O commands can be run.
Flash offset: 5F000 
D: Disable write protection
D: Erase offset: 389120, len: 64
D: Erase page 62 on bank 2
Reading 1 
D: Read offset: 389120, len: 64
Reading from flash .. 
ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
D: Write offset: 389120, len: 64
Reading 2 
D: Read offset: 389120, len: 64
Reading from flash .. 
00 01 02 03 04 05 06 07 08 09 0a 0b 0c 0d 0e 0f 10 11 12 13 14 15 16 17 18 19 1a 1b 1c 1d 1e 1f 20 21 22 23 24 25 26 27 28 29 2a 2b 2c 2d 2e 2f 30 31 32 33 34 35 36 37 38 39 3a 3b 3c 3d 3e 3f 
Erasing again... 
D: Erase offset: 389120, len: 64
D: Erase page 62 on bank 2
Reading 3 
D: Read offset: 389120, len: 64
Reading from flash .. 
00 01 02 03 04 05 06 07 08 09 0a 0b 0c 0d 0e 0f 10 11 12 13 14 15 16 17 18 19 1a 1b 1c 1d 1e 1f 20 21 22 23 24 25 26 27 28 29 2a 2b 2c 2d 2e 2f 30 31 32 33 34 35 36 37 38 39 3a 3b 3c 3d 3e 3f 
D: Enable write protection

azeemshatp commented:
> @azeemshatp,
> When you compile the app to run from bank2, can you tell me the value of switch CONFIG_FLASH_BASE_ADDRESS ?
> 0x8000000 or 0x8040000 ?

It was 0x8000000

ABOSTM commented:
> but the offset 0x50000 after remapping should correspond to the other bank. 

I don't catch your point: offset 0x50000 means address 0x08050000 (as CONFIG_FLASH_BASE_ADDRESS = 0x8000000) 
So it is in bank2 which is starting at 0x08050000 -->  (0x08050000 - 0x08050000) = 0x10000 --> page 32 of bank2
which correspond to the log:

> D: Erase page 32 on bank 2

azeemshatp commented:
> > but the offset 0x50000 after remapping should correspond to the other bank.
> 
> I don't catch your point: offset 0x50000 means address 0x08050000 (as CONFIG_FLASH_BASE_ADDRESS = 0x8000000)
> So it is in bank2 which is starting at 0x08050000 --> (0x08050000 - 0x08050000) = 0x10000 --> page 32 of bank2
> which correspond to the log:
> 
> > D: Erase page 32 on bank 2

As far as I understand based on how it worked in my previous board stm32l073rz, when the App runs from Bank 2 after setting the BFB2 bit, there is an address remapping happening in the background by the system bootloader.
If the case is similar for STM32G474, then after a bank switch, I think the address 0x08000000 becomes the base address of Bank 2 and for the Bank 1, the base address should be 0x08040000. 

So, If I want to write to the same physical flash location, while booting from either Bank 1 or Bank 2 (say 0x0801F800), then from Bank 1, the offset should be 0x01F800 and from Bank 2 the offset should be 0x5F800 (remapped to 0x0805F800).

azeemshatp commented:
Just got another observation now:
From Bank 2, using the below code, an intermediate flash erase to the offset 0x1F800 before erasing the offset 0x5F800 does some magic. 

offset = 0x1F800;
printf("\r\nIntermediate Erase to 0x%.X\r\n", offset);
flash_erase(flash_device, offset, size);
printf("\r\nIntermediate Reading from 0x%.X\r\n", offset);
ret = flash_read(flash_device, offset, buf, sizeof(buf));

It could then erase 0x5F800 properly as can be seen in the below log.

D: Flash initialized. BS: 8
D: Block 0: bs: 2048 count: 64
D: Block 1: bs: 131072 count: 1
D: Block 2: bs: 2048 count: 64
D: Disable write protection
*** Booting Zephyr OS build zephyr-v2.4.0  ***
CONFIG_FLASH_BASE_ADDRESS = 0x8000000 
Found flash controller FLASH_CTRL.
Flash I/O commands can be run.
Flash offset: 0x5F800 
D: Disable write protection
D: Erase offset: 391168, len: 64
D: Erase page 63 on bank 2

Reading 1 
D: Read offset: 391168, len: 64
Reading from flash .. 
ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
D: Write offset: 391168, len: 64

Reading 2 
D: Read offset: 391168, len: 64
Reading from flash .. 
00 01 02 03 04 05 06 07 08 09 0a 0b 0c 0d 0e 0f 10 11 12 13 14 15 16 17 18 19 1a 1b 1c 1d 1e 1f 20 21 22 23 24 25 26 27 28 29 2a 2b 2c 2d 2e 2f 30 31 32 33 34 35 36 37 38 39 3a 3b 3c 3d 3e 3f 

Intermediate Erase to 0x1F800
D: Erase offset: 129024, len: 64
D: Erase page 63 on bank 1

Intermediate Reading from 0x1F800
D: Read offset: 129024, len: 64
Reading from flash .. 
ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 

Erasing again at 0x5F800
D: Erase offset: 391168, len: 64
D: Erase page 63 on bank 2
Reading 3 
D: Read offset: 391168, len: 64
Reading from flash .. 
ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
D: Enable write protection
Success

ABOSTM commented:
This is strange.
Can you go a bit further and add a read at 0x5F800 just after intermediate erasing at 0x1F800 ?

azeemshatp commented:
That's interesting:
An erase at 0x1F800 seems to be resulting in an erase at both 0x1F800 and 0x5F800. But an erase at 0x5F800 doesn't seem to work at all.

	*** Booting Zephyr OS build zephyr-v2.4.0  ***
	CONFIG_FLASH_BASE_ADDRESS = 0x8000000 
	Found flash controller FLASH_CTRL.
	Flash I/O commands can be run.
	Flash offset: 0x5F800 
	D: Disable write protection
	D: Erase offset: 391168, len: 64
	D: Erase page 63 on bank 2

	Reading 1 
	D: Read offset: 391168, len: 64
	Reading from flash .. 
	ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
	D: Write offset: 391168, len: 64

	Reading 2 
	D: Read offset: 391168, len: 64
	Reading from flash .. 
	00 01 02 03 04 05 06 07 08 09 0a 0b 0c 0d 0e 0f 10 11 12 13 14 15 16 17 18 19 1a 1b 1c 1d 1e 1f 20 21 22 23 24 25 26 27 28 29 2a 2b 2c 2d 2e 2f 30 31 32 33 34 35 36 37 38 39 3a 3b 3c 3d 3e 3f 

	Intermediate Erase to 0x1F800
	D: Erase offset: 129024, len: 64
	D: Erase page 63 on bank 1

	Intermediate Reading from 0x1F800
	D: Read offset: 129024, len: 64
	Reading from flash .. 
	ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 

	Intermediate Reading from 0x5F800
	D: Read offset: 391168, len: 64
	Reading from flash .. 
	ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 

	Erasing again at 0x5F800
	D: Erase offset: 391168, len: 64
	D: Erase page 63 on bank 2
	Reading 3 
	D: Read offset: 391168, len: 64
	Reading from flash .. 
	ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
	D: Enable write protection
	Success

ABOSTM commented:
I think I got the point:
The driver is not designed to tackle bank swap.
This concerns only erase operation as it is necessary to select the bank (for read and write there is no bank selection).
Currently bank selection is based on offset: (offset > 0x40000) ==> select bank2 
But with a bank swap it is not valid. selection should be inverted.
For test purpose you can try to invert lines 119 and 124 of file drivers/flash/flash_stm32g4x.c 
I hope it works, but only for boot from bank2.
I will investigate further on monday how to make it work whatever bank is booted.

azeemshatp commented:
Ya, that seems to be the case. The driver doesn't handle the address remapping correctly. As you mentioned, inverting the lines 119 and 124 seem to work for Bank 2. Looking forward to an update next week. Thanks a lot

ABOSTM committed:
drivers: flash: flash_stm32g4x.c: take into account bank swap

When erasing flash page, take into account a potential bank swap.

Signed-off-by: Alexandre Bourdiol <alexandre.bourdiol@st.com>

ABOSTM commented:
Hi @azeemshatp thanks fro your feedback.
I pushed a 3rd commit thatI should fixe the swap issue.
I could not test it, as I still don't have proper bootloader revision (maybe tomorrow).
Can you please test it with your stm32g474rc chip ?

azeemshatp commented:
Hi @ABOSTM 
Thanks a lot for an early update on this today. 
I flashed the app to bank 2, and here is the response. I think there is still an issue with the page calculation while in Bank 2. 

Please see the wrong page sizes here: 
D: Erase page 191 on bank 1
D: Erase page 2097087 on bank 2

From Bank 1, everythings to be working as before.

	*** Booting Zephyr OS build zephyr-v2.4.0  ***
	CONFIG_FLASH_BASE_ADDRESS = 0x8000000 
	Found flash controller FLASH_CTRL.
	Flash I/O commands can be run.
	Flash offset: 0x5F800 
	D: Disable write protection
	D: Erase offset: 391168, len: 64
	D: Erase page 191 on bank 1

	Reading 1 
	D: Read offset: 391168, len: 64
	Reading from flash .. 
	ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
	D: Write offset: 391168, len: 64

	Reading 2 
	D: Read offset: 391168, len: 64
	Reading from flash .. 
	00 01 02 03 04 05 06 07 08 09 0a 0b 0c 0d 0e 0f 10 11 12 13 14 15 16 17 18 19 1a 1b 1c 1d 1e 1f 20 21 22 23 24 25 26 27 28 29 2a 2b 2c 2d 2e 2f 30 31 32 33 34 35 36 37 38 39 3a 3b 3c 3d 3e 3f 

	Erasing again at 0x5F800
	D: Erase offset: 391168, len: 64
	D: Erase page 191 on bank 1
	Reading 3 
	D: Read offset: 391168, len: 64
	Reading from flash .. 
	ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
	D: Enable write protection
	Flash offset: 0x1F800 
	D: Disable write protection
	D: Erase offset: 129024, len: 64
	D: Erase page 2097087 on bank 2
	D: Status: 0x00000080
	flash_erase failed (err:-5).

ABOSTM commented:
Sorry, I focused on bank selection, but miss to update page computation.
This is now done.

azeemshatp commented:
This is cool and everything seems to work fine now.
Tried to erase and write from both the Banks after bank switching.
Well done @ABOSTM and thanks a lot for fixing this.

ABOSTM commented:
@azeemshatp Thanks for your feedback,
I get a nucleo_g747re  with 0xFFFF00D4 bootloader version which supports bank swap.
I fake a smaller flash size, and it worked as expected.

@Laczen @nvlsianpu @martinjaeger @erwango  Can you please review ?

erwango commented on drivers/flash/flash_stm32g4x.c:
```diff
@@ -20,6 +20,9 @@ LOG_MODULE_REGISTER(LOG_DOMAIN);
 
 #define STM32G4X_PAGE_SHIFT	11
 
+#define STM32G4_FAMILY_MAX_FLASH	512
```
Nit: STM32 is the family, G4 is one series. So it would be `STM32G4_SERIES_MAX_FLASH`

ABOSTM replied:
Done

erwango commented on drivers/flash/flash_stm32g4x.c:
```diff
@@ -119,23 +113,24 @@ static int erase_page(const struct device *dev, unsigned int page)
-		}
-
-		/* The pages to be erased is in bank 2*/
+#if defined(FLASH_OPTR_DBANK)
```
You're assuming DBANK=1 if present.
If driver doesn't handle correctly case DBANK=0, that's ok for now.
But in this case, maybe we should we generate an error if this case is detected ?

ABOSTM replied:
NEw commit add for that.

erwango commented on drivers/flash/flash_stm32g4x.c:
```diff
@@ -101,11 +94,12 @@ static int write_dword(const struct device *dev, off_t offset, uint64_t val)
 }
 
-static int erase_page(const struct device *dev, unsigned int page)
+static int erase_page_offset(const struct device *dev, unsigned int offset)
```
I don't think the function name change was required as I am getting confused with the _offset suffix.
IMO, the change of parameter name was sufficient to notice the change if the function behavior.

ABOSTM replied:
Done

martinjaeger commented on drivers/flash/flash_stm32g4x.c:
```diff
@@ -28,17 +31,23 @@ bool flash_stm32_valid_range(const struct device *dev, off_t offset,
 {
+
+#if defined(FLASH_OPTR_DBANK) && (CONFIG_FLASH_SIZE < STM32G4_FAMILY_MAX_FLASH)
+	/*
+	 * In case of bank1/2 discontinuity range, is invalid in both cases:
+	 *  - offest is in bank1, and end of range is beyond bank1
+	 *  - offset is between bank1 and bank2
+	 */
+	if (((offset < FLASH_SIZE / 2) && (offset + len > FLASH_SIZE / 2))
+	    || ((offset >= FLASH_SIZE / 2) && (offset < BANK2_OFFSET))) {
+		return 0;
+	}
```
I think the number of comparisons could be reduced here while still getting the same result, but not sure if that makes it harder to read/understand. Up to you to decide.
```suggestion
	/*
	 * In case of bank1/2 discontinuity, the range should not end
	 * beyond bank1 and start before bank2 at the same time.
	 * Locations beyond bank2 are caught by flash_stm32_range_exists.
	 */
	if ((offset + len > FLASH_SIZE / 2) && (offset < BANK2_OFFSET)) {
		return 0;
	}
```

martinjaeger replied:
Consider this a comment only and not a -1. Will approve after erwango's comments are addressed. And I will also give it another try in hardware in the next days.

ABOSTM replied:
Done

martinjaeger commented on drivers/flash/flash_stm32g4x.c:
```diff
@@ -181,18 +205,34 @@ void flash_stm32_page_layout(const struct device *dev,
 
-	*layout = &stm32g4_flash_layout;
-	*layout_size = 1;
+	*layout = stm32g4_flash_layout;
```
The `&` has to be kept here because `layout` is a pointer to a pointer.
```suggestion
	*layout = &stm32g4_flash_layout;
```

martinjaeger replied:
Oops, forget about this. It's too late... we have an array now, so this is totally correct. Otherwise would have been strange also that it worked in Azeemsha's test.

ABOSTM committed:
divers: flash: flash_stm32: add flash configuration check

Add weak function to check flash configuration.
On stm32g4: single bank configuration not supported
when dual bank capable.

Signed-off-by: Alexandre Bourdiol <alexandre.bourdiol@st.com>

ABOSTM commented on drivers/flash/flash_stm32g4x.c:
```diff
@@ -48,13 +48,6 @@ bool flash_stm32_valid_range(const struct device *dev, off_t offset,
- */
-static unsigned int get_page(off_t offset)
-{
-	return offset >> STM32G4X_PAGE_SHIFT;
```
Yes, done

martinjaeger reviewed (approved):
I can confirm that it works perfectly fine in hardware with STM32G474RCT6.

Tested the flash erase + write + read operations from both banks at the following offsets:

- `0x1F800` (own bank) works
- `0x20000` (beyond own bank) fails, as it should
- `0x40000` (beginning of other bank) works
- `0x5F800` (end of other bank) works
- `0x60000` (beyond other bank) fails, as it should

Laczen reviewed (approved):
Nice work !

erwango commented on drivers/flash/flash_stm32.c:
```diff
@@ -383,6 +389,9 @@ static int stm32_flash_init(const struct device *dev)
 		flash_stm32_parameters.write_block_size);
 
+	/* Check Flash configuration */
+	flash_stm32_check_configuration();
```
Shouldn't we check function's output here?

ABOSTM replied:
Yes, indeed.
Done
